### PR TITLE
Snake Plugin: Store game state on close and restore it on restart, show highscore

### PR DESCRIPTION
**Pull request by JuanJakobo**
# What's new

- Implements #1041 as it safes the current game state and restores it on restart
- Saves also an highscore to an file and shows the highscore on GameOver

# Verification 

- Open App and close it after moving a little bit around, reopen it and the state should be as before. 
- Close the App when GameOver Screen is shown. The Game should restart normal.
- Crash the snake and see if the highscore is shown. Do it again with an higher score and see if it increased, further do it with a lower score to verify if it stays the same.

# Checklist (For Reviewer)

- [ ] PR has description of feature/bug or link to Confluence/Jira task
- [ ] Description contains actions to verify feature/bugfix
- [ ] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

**Comment by skotopes** [comment]
@DrZlo13 is currently formalizing rules for ext apps storage layout. He'll be reviewer for this PR

**Comment by DrZlo13** [comment]
@JuanJakobo Hello!
I am now busy formalizing the rules for placing user files and supporting these rules in the firmware. It's enough for me for now if you put this file in "/ext/apps_data/snake_game".

**Comment by JuanJakobo** [comment]
@DrZlo13 Hi, 
I moved the file storage to "/ext/apps_data/snake_game".

**Comment by skotopes** [comment]
@DrZlo13 please review.

File: applications/plugins/snake_game/helpers/snake_file_handler.c [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+    if(storage_common_stat(storage, SNAKE_GAME_FILE_PATH, NULL) == FSE_OK) {
+        if(!flipper_format_file_open_existing(file, SNAKE_GAME_FILE_PATH)) {
+            snake_game_close_file(file);
+            return NULL;
+        }
+    } else {
+        if(storage_common_stat(storage, APPS_DATA, NULL) == FSE_NOT_EXIST) {
+            if(!storage_simply_mkdir(storage, APPS_DATA)) {
+                return NULL;
```
Comment (skotopes) [inline applications/plugins/snake_game/helpers/snake_file_handler.c:28]: Going to leak file if something will go wrong here.
Whole way of working with file looks quite dangerous, please check how it's done in other apps.

File: applications/plugins/snake_game/helpers/snake_file_handler.c [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+        furi_record_close(RECORD_STORAGE);
+        return;
+    }
+    flipper_format_file_close(file);
+    flipper_format_free(file);
+    furi_record_close(RECORD_STORAGE);
+}
+
+static FlipperFormat* snake_game_open_file() {
+    Storage* storage = furi_record_open(RECORD_STORAGE);
```
Comment (skotopes) [inline applications/plugins/snake_game/helpers/snake_file_handler.c:17]: Opening and closing storage record doesn't looks good here. The best option is to do it in allocator/deallocator of snake app itself and then pass it here.

File: applications/plugins/snake_game/helpers/snake_file_handler.c [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+        }
+    } else {
+        if(storage_common_stat(storage, APPS_DATA, NULL) == FSE_NOT_EXIST) {
+            if(!storage_simply_mkdir(storage, APPS_DATA)) {
+                return NULL;
+            }
+        }
+        if(storage_common_stat(storage, SNAKE_GAME_FILE_DIR_PATH, NULL) == FSE_NOT_EXIST) {
+            if(!storage_simply_mkdir(storage, SNAKE_GAME_FILE_DIR_PATH)) {
+                return NULL;
```
Comment (skotopes) [inline applications/plugins/snake_game/helpers/snake_file_handler.c:33]: Leak here too

File: applications/plugins/snake_game/helpers/snake_file_handler.c [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+bool snake_game_init_game_from_file(SnakeState* const snake_state) {
+    FlipperFormat* file = snake_game_open_file();
+
+    if(file != NULL) {
+        FuriString* file_type = furi_string_alloc();
+        uint32_t version = 1;
+        if(!flipper_format_read_header(file, file_type, &version)) {
+            furi_string_free(file_type);
+            snake_game_close_file(file);
+            return false;
```
Comment (skotopes) [inline applications/plugins/snake_game/helpers/snake_file_handler.c:120]: We usually do {} while(0) and break. Please check how other apps do that

**Review by skotopes — CHANGES_REQUESTED** [review]

**Comment by skotopes** [comment]
Please un-draft when ready ;-) And let me know if you need any help

**Comment by fennectech** [comment]
I can't tell whether this means it was merged or is dead.

**Comment by skotopes** [comment]
@fennectech dead, PR was closed before merged

**Pull request closed without merging**